File: sprint_3/misc/selection_sort.c
/*
Insertion sort.
Time: O(n**2). Space: O(1)
*/

#include <stdio.h>

#define ARR_LEN 5

void selection_sort(int *array, size_t len)
{
    for (int i = 0; i < len; i++)
    {
        int elem_to_insert = array[i];
        int j = i;
        while (j > 0 && array[j] < array[j - 1])
        {
            array[j] = array[j - 1];
            j--;
        }
        array[j] = elem_to_insert;
    }
}

void print_array(int *array, int len)
{
    for (int i = 0; i < len; i++)
    {
        printf("%d ", array[i]);
    }
    printf(" \n");
}

int main()
{
    int arr[ARR_LEN] = {-1, -8, 0, 44, 12};
    puts("array before:");
    print_array(arr, ARR_LEN);
    selection_sort(arr, ARR_LEN);
    puts("array after:");
    print_array(arr, ARR_LEN);
    return 0;
}

File: sprint_3/misc/selection_sort.py
"""
Insertion sort.
Time: O(n**2). Space: O(1)
"""

from copy import deepcopy
from dataclasses import dataclass


def selection_sort(arr: list[int]) -> None:
    for i in range(1, len(arr)):
        item_to_insert = arr[i]
        j = i
        while j > 0 and item_to_insert < arr[j - 1]:
            arr[j] = arr[j - 1]
            j -= 1
        arr[j] = item_to_insert
        print(f"step {i}, sorted {i + 1} elements, array {arr}")


@dataclass
class TestData:
    arr: list[int]
    expected: list[int]


def test():
    tds = [
        TestData([], []),
        TestData([-1, -4, 0, -2], [-4, -2, -1, 0]),
        TestData([89, 4, 11, -8, 14], [-8, 4, 11, 14, 89]),
        TestData([4, 89, 11, -8, 14, -22, -8, 0], [-22, -8, -8, 0, 4, 11, 14, 89]),
    ]
    for td in tds:
        original_arr = deepcopy(td.arr)
        selection_sort(td.arr)
        assert td.arr == td.expected, f"expected {td.expected}, got {td.arr}, input: {original_arr}"


if __name__ == "__main__":
    test()
